[GPU Process] RemoteRenderingBackend must ignore resource messages once m_remoteResourceCache is cleared

shallawa opened the pull request:
#### 682daa7855096d0abf3c178d4d139b0f2f99c689
<pre>
[GPU Process] RemoteRenderingBackend must ignore resource messages once m_remoteResourceCache is cleared
<a href="https://bugs.webkit.org/show_bug.cgi?id=242142">https://bugs.webkit.org/show_bug.cgi?id=242142</a>
rdar://94143906

Reviewed by Chris Dumez.

RemoteRenderingBackend::stopListeningForIPC() is calling StreamServerConnection
::stopReceivingMessages() after it dispatches a code block to clear its
m_remoteResourceCache. But incoming messages can still be dispatched to
RemoteRenderingBackend between executing the clearing block and and stopping
receiving messages.

If the RemoteRenderingBackend::ReleaseRemoteResource message is received after
clearing m_remoteResourceCache, GPUProcess will simulate-crash with the message:
&quot;Resource is being released before being cached.&quot;

The fix is to ensure that the clearing block is executed the last thing on the
WorkQueue after processing all the queued messages and before stopping receiving
new messages. This can be done by passing the clearing block to stopReceivingMessages()
which is going to be executed when m_shouldQuit is true and calling processStreams()
in StreamConnectionWorkQueue::startProcessingThread().

Remove RemoteRenderingBackend::m_renderingResourcesRequest and
updateSharedMemoryForGetPixelBufferHelper() since they are not used.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::stopListeningForIPC):

Canonical link: <a href="https://commits.webkit.org/252187@main">https://commits.webkit.org/252187@main</a>
</pre>

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -117,7 +117,9 @@ void RemoteRenderingBackend::startListeningForIPC()
 {
     ASSERT(RunLoop::isMain());
+
     m_streamConnection->invalidate();
```
The idea is that this call stops the IPC messages.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -117,7 +117,9 @@ void RemoteRenderingBackend::startListeningForIPC()
+    m_streamConnection->stopReceivingMessages(Messages::RemoteRenderingBackend::messageReceiverName(), m_renderingBackendIdentifier.toUInt64());
 
     // This item is dispatched to the WorkQueue before calling stopAndWaitForCompletion() such that it will process it last, after any existing work.
     m_workQueue->dispatch([&] {
```
The idea is that this task is the last task that is executed.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -117,7 +117,9 @@ void RemoteRenderingBackend::startListeningForIPC()
     ASSERT(RunLoop::isMain());
+
     m_streamConnection->invalidate();
+    m_streamConnection->stopReceivingMessages(Messages::RemoteRenderingBackend::messageReceiverName(), m_renderingBackendIdentifier.toUInt64());
```
if the `invalidate()` does not stop the processing of the messages, then `stopReceivingMessages` would be needed.
However, that would mean that if `stopReceivingMessages` would be effective, then there would be a message being dispatched that did not have a receiver. 
This causes an assertion.

kkinnunen-apple reviewed (commented):
The issue of moving the `stopReceivingMessages` up is that it doesn't matter whether or not it's before the cleanup task submission -- there's no ordering guarantee either way.

The work loop is:
1.(w) get connections and functions
2.(w) run functions
3.(w) dispatch connections

Your call order might be:

1. (c) stopReceivingMessages
2. (c) dispatch(cleanup)

Suppose the RRB has done 2.(c). If the work queue is between 1(w) and 2(w) , it still runs the cleanup before dispatching the messages in the connections.


The idea of the existing code is:

1. streamConnection->invalidate() ---  Signal that new messages should not be be dispatched (note: it doesn't stop right away, as the messages are dispatched in other thread.
5. workQueue->dispatch(cleanup task) -- Schedule a cleanup task to be run after the above signal reaches the work queue
6. wait until cleanup task has run
7. remove the message receiver for book keeping reasons (should not affect the runtime behaviour)

Would you be able to fix this as follows:

1. std::atomic<bool> StreamServerConnection::m_isValid
4. toggle isValid = false in StreamServerConnection::invalidate()
5. check the m_isValid like so:


```
StreamServerConnection::DispatchResult StreamServerConnection::dispatchStreamMessages(size_t messageLimit)
{
    RefPtr<StreamMessageReceiver> currentReceiver;
    // FIXME: Implement WTF::isValid(ReceiverName).
    uint8_t currentReceiverName = static_cast<uint8_t>(ReceiverName::Invalid);

    for (size_t i = 0; i < messageLimit && m_isValid; ++i) {
```

This way the invalidate will stop the connection and the cleanup task will be run after the last  dispatched message.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -374,7 +374,7 @@ void RemoteRenderingBackend::releaseRemoteResourceWithQualifiedIdentifier(Qualif
     }
     auto success = m_remoteResourceCache.releaseRemoteResource(renderingResourceIdentifier);
-    MESSAGE_CHECK(success, "Resource is being released before being cached.");
+    MESSAGE_CHECK(success || !m_streamConnection->connection().isValid(), "Resource is being released before being cached.");
```
This is not correct.
The message can be processed at arbitrary time.
The connection may be invalidated at arbitrary time.

It is valid to dispatch the "release remote resource" message and get the connection invalidated during the dispatch.

shallawa replied:
I changed m_remoteResourceCache to be a unique_ptr. I set it to null in the clearing block and I check it before using it. So adding to it or removing from it will not happen once it is cleared.

Clearing, checking and accessing the m_remoteResourceCache happen only on the WorkQueue. So there should not be a problem due to ordering or synchronization here.

cdumez commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
```diff
@@ -79,14 +79,15 @@ void StreamConnectionWorkQueue::removeStreamConnection(StreamServerConnection& c
 }
 
-void StreamConnectionWorkQueue::stopAndWaitForCompletion()
+void StreamConnectionWorkQueue::stopAndWaitForCompletion(std::optional<WTF::Function<void()>>&& cleanup)
```
A WTF::Function can already be null so we shouldn't wrap it in an optional.

Just use WTF::Function<void()>&&

smfr replied:
Maybe call the argument cleanupFunction or postCompletionFunction

cdumez commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.h:
```diff
@@ -69,6 +69,7 @@ class WTF_CAPABILITY("is current") StreamConnectionWorkQueue final : public Func
     mutable Lock m_lock;
     RefPtr<Thread> m_processingThread WTF_GUARDED_BY_LOCK(m_lock);
     Deque<Function<void()>> m_functions WTF_GUARDED_BY_LOCK(m_lock);
+    std::optional<WTF::Function<void()>> m_cleanup WTF_GUARDED_BY_LOCK(m_lock);
```
Should just be `WTF::Function<void()>`, not need to wrap in an optional. a Function has a null state and can be null-checked.

cdumez replied:
Can we call this m_cleanupFunction or something similar? Just m_cleanup doesn't seem great.

smfr commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.h:
```diff
@@ -69,6 +69,7 @@ class WTF_CAPABILITY("is current") StreamConnectionWorkQueue final : public Func
     mutable Lock m_lock;
     RefPtr<Thread> m_processingThread WTF_GUARDED_BY_LOCK(m_lock);
     Deque<Function<void()>> m_functions WTF_GUARDED_BY_LOCK(m_lock);
+    std::optional<WTF::Function<void()>> m_cleanup WTF_GUARDED_BY_LOCK(m_lock);
```
Maybe call this m_cleanupFunction or m_postCompletionFunction

cdumez commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
```diff
@@ -111,6 +112,11 @@ void StreamConnectionWorkQueue::startProcessingThread()
+                {
+                    Locker locker { m_lock };
+                        if (m_cleanup)
+                            (*m_cleanup)();
```
m_cleanupFunction();

cdumez commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.h:
```diff
@@ -49,7 +49,7 @@ class WTF_CAPABILITY("is current") StreamConnectionWorkQueue final : public Func
 
     void dispatch(WTF::Function<void()>&&) final;
-    void stopAndWaitForCompletion();
+    void stopAndWaitForCompletion(std::optional<WTF::Function<void()>>&& cleanup = std::nullopt);
```
void stopAndWaitForCompletion(WTF::Function<void()>&& cleanup = nullptr);

cdumez commented on Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
```diff
@@ -111,6 +112,11 @@ void StreamConnectionWorkQueue::startProcessingThread()
                 processStreams();
+                {
+                    Locker locker { m_lock };
+                        if (m_cleanup)
```
indentation seems wrong.

cdumez reviewed (approved):
LGTM with fixes.

shallawa committed:
[GPU Process] RemoteRenderingBackend must ignore resource messages once m_remoteResourceCache is cleared
https://bugs.webkit.org/show_bug.cgi?id=242142
rdar://94143906

Reviewed by Chris Dumez.

RemoteRenderingBackend::stopListeningForIPC() is calling StreamServerConnection
::stopReceivingMessages() after it dispatches a code block to clear its
m_remoteResourceCache. But incoming messages can still be dispatched to
RemoteRenderingBackend between executing the clearing block and and stopping
receiving messages.

If the RemoteRenderingBackend::ReleaseRemoteResource message is received after
clearing m_remoteResourceCache, GPUProcess will simulate-crash with the message:
"Resource is being released before being cached."

The fix is to ensure that the clearing block is executed the last thing on the
WorkQueue after processing all the queued messages and before stopping receiving
new messages. This can be done by passing the clearing block to stopReceivingMessages()
which is going to be executed when m_shouldQuit is true and calling processStreams()
in StreamConnectionWorkQueue::startProcessingThread().

Remove RemoteRenderingBackend::m_renderingResourcesRequest and
updateSharedMemoryForGetPixelBufferHelper() since they are not used.

* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::stopListeningForIPC):

Canonical link: https://commits.webkit.org/252187@main
